File: client/src/app/models/closure-schedule.ts
export type ClosureScheduleType = 'ANNUAL' | 'PERMANENT' | 'FIXED';

export class ClosureSchedule {
  id: string | null = null;
  scheduleType: ClosureScheduleType;
  reason: string | null = null;
  startDate: string | null = null;
  endDate: string | null = null;
  startMonth: number | null = null;
  startDay: number | null = null;
  endMonth: number | null = null;
  endDay: number | null = null;

  public static deserialize(payload: any): ClosureSchedule {
    const schedule = new ClosureSchedule();
    schedule.id = payload.id ?? null;
    schedule.scheduleType = payload.scheduleType;
    schedule.reason = payload.reason ?? null;
    schedule.startDate = payload.startDate ?? null;
    schedule.endDate = payload.endDate ?? null;
    schedule.startMonth = payload.startMonth ?? null;
    schedule.startDay = payload.startDay ?? null;
    schedule.endMonth = payload.endMonth ?? null;
    schedule.endDay = payload.endDay ?? null;
    return schedule;
  }

  public static serialize(schedule: ClosureSchedule): any {
    const payload: Record<string, unknown> = {
      id: schedule.id,
      scheduleType: schedule.scheduleType,
      reason: schedule.reason,
    };
    if (schedule.scheduleType === 'ANNUAL') {
      payload.startMonth = schedule.startMonth;
      payload.startDay = schedule.startDay;
      payload.endMonth = schedule.endMonth;
      payload.endDay = schedule.endDay;
    }
    if (schedule.scheduleType === 'FIXED') {
      payload.startDate = schedule.startDate;
      payload.endDate = schedule.endDate;
    }
    return payload;
  }
}

File: server/src/models/closure_schedule.py
from sqlalchemy.dialects.postgresql import UUID

from extensions import db
from models.base_entity import BaseEntity
from models.enums.closure_schedule_type_enum import ClosureScheduleTypeEnum


class ClosureSchedule(BaseEntity):
    """A single closure window attached to one closable topo entity."""

    __tablename__ = "closure_schedules"

    schedule_type = db.Column(db.Enum(ClosureScheduleTypeEnum), nullable=False)
    reason = db.Column(db.Text, nullable=True)

    start_date = db.Column(db.Date, nullable=True)
    end_date = db.Column(db.Date, nullable=True)
    start_month = db.Column(db.SmallInteger, nullable=True)
    start_day = db.Column(db.SmallInteger, nullable=True)
    end_month = db.Column(db.SmallInteger, nullable=True)
    end_day = db.Column(db.SmallInteger, nullable=True)

    crag_id = db.Column(UUID(), db.ForeignKey("crags.id", ondelete="CASCADE"), nullable=True)
    sector_id = db.Column(UUID(), db.ForeignKey("sectors.id", ondelete="CASCADE"), nullable=True)
    area_id = db.Column(UUID(), db.ForeignKey("areas.id", ondelete="CASCADE"), nullable=True)
    line_id = db.Column(UUID(), db.ForeignKey("lines.id", ondelete="CASCADE"), nullable=True)

    crag = db.relationship(
        "Crag",
        backref=db.backref("closure_schedules", lazy="select", cascade="all, delete-orphan"),
    )
    sector = db.relationship(
        "Sector",
        backref=db.backref("closure_schedules", lazy="select", cascade="all, delete-orphan"),
    )
    area = db.relationship(
        "Area",
        backref=db.backref("closure_schedules", lazy="select", cascade="all, delete-orphan"),
    )
    line = db.relationship(
        "Line",
        backref=db.backref("closure_schedules", lazy="select", cascade="all, delete-orphan"),
    )
